File: frontend/pages/api/renameFile.ts
// pages/api/renameFile.ts
import { NextApiRequest, NextApiResponse } from "next";
import fs from "fs/promises";
import path from "path";
import simpleGit from "simple-git";

const git = simpleGit();

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  if (req.method !== "POST") return res.status(405).end();

  const { oldUri, newUri } = req.body;
  const root = path.resolve(process.cwd(), "savedFiles");

  const oldPath = path.join(root, path.basename(oldUri));
  const newPath = path.join(root, path.basename(newUri));

  try {
    await fs.rename(oldPath, newPath);
    await git.mv(oldPath, newPath);
    await git.commit(
      `Renamed file: ${path.basename(oldUri)} → ${path.basename(newUri)}`,
    );
    return res.status(200).json({ success: true });
  } catch (err) {
    console.error("Rename error:", err);
    return res.status(500).json({ success: false, message: err.message });
  }
}

File: app/routes/rename_file.py
from flask import Blueprint, request, jsonify
import os
import subprocess
import shutil

rename_file_bp = Blueprint("rename_file", __name__)

METADATA_DIR = "savedFiles/.metadata"

def uri_to_path(uri: str) -> str:
    if uri.startswith("file://"):
        return uri[7:]
    return uri

def git_rename(old_path: str, new_path: str):
    subprocess.run(["git", "mv", old_path, new_path], check=True)
    subprocess.run(["git", "add", new_path], check=True)

def rename_metadata(old_uri: str, new_uri: str):
    old_base = os.path.splitext(os.path.basename(uri_to_path(old_uri)))[0]
    new_base = os.path.splitext(os.path.basename(uri_to_path(new_uri)))[0]
    old_meta_path = os.path.join(METADATA_DIR, f"{old_base}.metadata.json")
    new_meta_path = os.path.join(METADATA_DIR, f"{new_base}.metadata.json")
    if os.path.exists(old_meta_path):
        os.makedirs(METADATA_DIR, exist_ok=True)
        shutil.move(old_meta_path, new_meta_path)

@rename_file_bp.route("/api/renameFile", methods=["POST"])
def rename_file():
    data = request.json
    old_uri = data.get("oldUri")
    new_uri = data.get("newUri")
    if not old_uri or not new_uri:
        return jsonify({"success": False, "message": "Missing oldUri or newUri"}), 400

    try:
        old_path = uri_to_path(old_uri)
        new_path = uri_to_path(new_uri)

        os.rename(old_path, new_path)
        git_rename(old_path, new_path)
        rename_metadata(old_uri, new_uri)

        return jsonify({"success": True})
    except Exception as e:
        print(f"Error in rename_file: {e}")
        return jsonify({"success": False, "message": str(e)}), 500
